Adds instrumentation for search path

Gaganjuneja opened the pull request:
### Description
Adds the instrumentation on Search path. Making it a Draft PR to discuss the instrumentation on SearchService.java whether this is the right way to do explicit tracing or should we do inside SearchOperationsListener, which will require the state management.

### Related Issues
Resolves #[7546]


### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Gaganjuneja commented:
@reta @shwetathareja Please provide your thoughts on this.

Gaganjuneja commented:
@reta @shwetathareja I have also merged the #8411 changes. Please take a look.

shwetathareja commented on server/src/main/java/org/opensearch/action/search/TransportSearchAction.java:
```diff
@@ -286,7 +292,9 @@ protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<
         }
-        executeRequest(task, searchRequest, this::searchAsyncAction, listener);
+        SpanScope scope = tracerFactory.getTracer().startSpan("SearchTask_" + task.getId());
+        TracingActionListener tracingActionListener = new TracingActionListener(tracerFactory, listener, scope);
```
TracingAwareListenerWrapper. 
I am not an expert on search code path but do you think instrumenting ```SearchOperationListener``` would easy / clean access to different phases?

Gaganjuneja replied:
@shwetathareja I tried implementing SearchOperationListener for instrumenting query and fetch phase but since now we are returning the SpanScope from the startSpan, hence we need to maintain the state (SpanScope returned from the preQueryPhase needs to be cached for ending in the onQueryPhase method.). We can do that in the ThreadLocal since this runs in a single threaded fashion as far as I understood. Need your thoughts on that? @reta

Gaganjuneja commented on server/src/main/java/org/opensearch/search/SearchService.java:
```diff
@@ -590,13 +595,18 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
                 afterQueryTime = executor.success();
+            } catch (Exception e) {
+                spanScope.setError(e);
+                afterQueryTime = 0l;
```
Ah, i missed pushing the latest change, this we did just for testing purposes.

Gaganjuneja commented on server/src/main/java/org/opensearch/search/SearchService.java:
```diff
@@ -621,14 +631,23 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
             }
             executor.success();
+        } catch (Exception e) {
+            spanScope.setError(e);
```
done

Gaganjuneja commented on server/src/main/java/org/opensearch/action/search/TransportSearchAction.java:
```diff
@@ -286,7 +292,9 @@ protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<
-        executeRequest(task, searchRequest, this::searchAsyncAction, listener);
+        SpanScope scope = tracerFactory.getTracer().startSpan("SearchTask_" + task.getId());
+        TracingActionListener tracingActionListener = new TracingActionListener(tracerFactory, listener, scope);
+        executeRequest(task, searchRequest, this::searchAsyncAction, tracingActionListener);
```
@shwetathareja, Here are the sample spans.
```
QueryPhase_[my-index-0000021][0]' : e79d8aa8d6d81b267283f06c084c05e1 57af1b95e8412a91 INTERNAL [tracer: os-tracer:] AttributesMap{data={shard_id=0, th_name=opensearch[runTask-0][search][T#3], node_id=7ZhAHn8OSx6dVMWVdxNfpw}, capacity=128, totalAddedValues=3}
```
```
'FetchPhase_[my-index-000002][0]' : e79d8aa8d6d81b267283f06c084c05e1 f8aa81e7496b3fc6 INTERNAL [tracer: os-tracer:] AttributesMap{data={shard_id=0, th_name=opensearch[runTask-0][search][T#13], node_id=7ZhAHn8OSx6dVMWVdxNfpw}, capacity=128, totalAddedValues=3}
```

Yes, we will add instrumentations in the foundational classes like RestAction, TransportAction and TaskManager etc.

Gaganjuneja commented:
Thanks @reta for reviewing. I really apologize for not setting the proper context. Actual plan is like you proposed. But there are a couple of issues at this point in time.
1. In the 2.9 experimental release, I just wanted to provide some basic instrumentation on one of the operations paths so that users can have some example and I chose search for that.
2. During initial prototyping I added instrumentations in foundational classes only like you proposed. But it generates a lot of spans and since we are not done with that level of performance testing and lacking the integrations with other OTEL exporters (though I am planning to raise that PR in a while) so it would be difficult for anyone to consume through logs. Still we can do the instrumentation for search path only in the foundational classes but it requires some search specific dirty checks like `resource should contain _search`, `instance of SearchTask, ShardSearchTask`, etc.
3. With these recent telemetry framework changes now we need to pass the SpanScope between multiple classes/methods which touches most of the foundational piece so will do that in the next release.
4. I tried adding the instrumentation at the coordinator as well but facing issue where we are messing up with setting parent when multiple tasks are being submitted from a single thread, opened a issue #8428

So, the only motivation behind this instrumentation is to provide one sample so that users/developers can refer to it and start adding the spans wherever needed.

your thoughts?

Gaganjuneja commented:
@reta , I agree with you. But one point I would like to discuss. There are so many transport actions such as node stats, background jobs, search, indexing, health checks, etc. If we add instrumentation generically at these code paths then there will be a flood of spans. We might need to introduce the action specific knobs to control this. Otherwise we can start iteratively action by action. But in both of the cases we can add support at the rest layer level to intercept the X-Opaque-Id and make it available for downstream instrumentation.

reta commented:
> If we add instrumentation generically at these code paths then there will be a flood of spans.

That's the whole purpose of the tracing - observe the whole system. You don't need any knobs on span level:
 - you have samplers (to limit when trace)
 - you have flags (we don't use them now but we will I think) to limit when to trace
 - and you have on/off setting

> But in both of the cases we can add support at the rest layer level to intercept the X-Opaque-Id and make it available for downstream instrumentation.

We don't need X-Opaque-Id - I referred that as an example. What we need is the trace span in headers coming from the ingress (client / mesh / gateway / you name it)

Gaganjuneja commented:
> * you have flags (we don't use them now but we will I think) to limit when to trace

By flag you mean levels or some component level settings?

> We don't need X-Opaque-Id - I referred that as an example of the header being propagated. What we need is the trace span in headers coming from the ingress (client / mesh / gateway / you name it)

Yes, I also meant the same.

reta commented:
> By flag you mean levels or some component level settings?

Nope, see please [1]

    Trace Flags, a binary encoding containing information about the trace

[1] https://opentelemetry.io/docs/concepts/signals/traces/#span-context

Gaganjuneja commented:
@reta , I am working on documentation [task](https://github.com/opensearch-project/documentation-website/issues/3951) for tracing framework features for [release 2.9](https://github.com/orgs/opensearch-project/projects/1). Which includes how somebody can enable/disable the feature and other settings. While I fully appreciate and support the holistic approach you've outlined for adding instrumentation, I believe it would be beneficial to include a basic search request instrumentation as a stepping stone that users can easily relate to. (SearchTask -> QueryPhase -> FetchPhase)

I have created a [meta task](https://github.com/opensearch-project/OpenSearch/issues/8557) for the overall tracing implementation and will continue working on that. I tried doing the changes for TransportAction but it's a lot of work and requires testing as there are hundreds of implementations of TransportAction and we need to pass the [TracingFactory](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/telemetry/tracing/TracerFactory.java) instance from all concrete classes.

reta commented:
> While I fully appreciate and support the holistic approach you've outlined for adding instrumentation, I believe it would be beneficial to include a basic search request instrumentation as a stepping stone that users can easily relate to. (SearchTask -> QueryPhase -> FetchPhase)

As I mentioned, I don't see the point in rushing the instrumentation in the current form (it could be replaced by a simple log statements if there is an urgent need to do so). If others see the value in it - I have no objections.

Bukhtawar commented:
Thanks @Gaganjuneja for this, wanted to understand more around the changes. As I see, the tracing is being done mostly around the shard level search execution `SearchServices` as long as changes are pertaining to this class or specific search operations, without touching the distributed transport layers and task coordinations I would still be fine with it since highly likely they won't change. But the moment you change the transport layers `TransportSearchAction` it makes me concerned if we are writing it generic enough and thinking through how the framework interactions would evolve. 
If you can isolate core search module instrumentations and still use this PR to build search shard level interactions more generically, I am good.

reta commented:
Thanks @Bukhtawar 

> @reta what do you think?

We already found one with Tracing API [1], this is good (in my opinion) and we could address that now. 

> But the moment you change the transport layers TransportSearchAction it makes me concerned if we are writing it generic enough and thinking through how the framework interactions would evolve.

That the most complicated and arguably the most important part of the work. It opens up the way for context transfer and actually getting real benefits of distributed nature of the tracing (not only for core but beyond that). What we get with this (isolated) instrumentation - out of the blue uncorrelated traces. This is what logs are for.

shwetathareja commented:
@Gaganjuneja @Bukhtawar 
I completely agree to that point that we should have more sample instrumentation across different code paths to see if there are gaps with the tracing framework we have built so far. But agree with @reta here that thinking on how we are going to introduce the tracing at the transport level generically is important as without that others might not be able to use it holistically. 
This PR does start the discussion on one other aspect whether we should simply do inline instrumentation vs use the current listeners ```SearchOperationListener``` and what changes we need for the same.

Gaganjuneja commented on server/src/main/java/org/opensearch/search/SearchService.java:
```diff
@@ -590,13 +595,18 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
-            final long afterQueryTime;
-            try (SearchOperationListenerExecutor executor = new SearchOperationListenerExecutor(context)) {
+            long afterQueryTime;
+            final SpanScope spanScope = tracerFactory.getTracer().startSpan("QueryPhase_" + context.shardTarget().getShardId());
```
Opened a bug #8561 to address this.

suranjay replied:
@reta Raised a [PR](https://github.com/opensearch-project/OpenSearch/pull/8565) for this.

suranjay replied:
@reta With this change, shall we use `Tracer.java` instead `TracerFactory.java` in instrumentation code?

Gaganjuneja commented:
> @Gaganjuneja @Bukhtawar I completely agree to that point that we should have more sample instrumentation across different code paths to see if there are gaps with the tracing framework we have built so far. But agree with @reta here that thinking on how we are going to introduce the tracing at the transport level generically is important as without that others might not be able to use it holistically. This PR does start the discussion on one other aspect whether we should simply do inline instrumentation vs use the current listeners `SearchOperationListener` and what changes we need for the same.

Absolutely, @shwetathareja, We need to decide on that as well, explicit inline instrumentation looks good to me.

Gaganjuneja commented:
> Thanks @Bukhtawar
> 
> > @reta what do you think?
> 
> We already found one with Tracing API [1], this is good (in my opinion) and we could address that now.
> 
> > But the moment you change the transport layers TransportSearchAction it makes me concerned if we are writing it generic enough and thinking through how the framework interactions would evolve.
> 
> That the most complicated and arguably the most important part of the work. It opens up the way for context transfer and actually getting real benefits of distributed nature of the tracing (not only for core but beyond that). What we get with this (isolated) instrumentation - out of the blue uncorrelated traces. This is what logs are for (my opinion).

@reta @Bukhtawar Please take a look once. I have added the instrumentation in TransportAction directly and removed from the TransportSearchAction.

kotwanikunal commented:
Apologies. This PR was auto closed without reaching a resolution from the maintainers. 
Re-opening to move it forward.
Thanks for your contributions to OpenSearch!

kotwanikunal commented:
There has been no movement on this PR. @Gaganjuneja Please reopen if you still plan to work on it.
